homeassistant/components/tplink_omada/switch.py

```diff
@@ -51,37 +58,104 @@ async def async_setup_entry(
         coordinator = controller.get_switch_port_coordinator(switch)
         await coordinator.async_request_refresh()
 
-        for idx, port_id in enumerate(coordinator.data):
-            if idx < switch.device_capabilities.poe_ports:
-                entities.append(
-                    OmadaNetworkSwitchPortPoEControl(coordinator, switch, port_id)
-                )
+        entities.extend(
+            OmadaDevicePortSwitchEntity[
+                OmadaSwitchPortCoordinator, OmadaSwitch, OmadaSwitchPortDetails
+            ](coordinator, switch, port.port_id, _get_switch_port_base_name(port), desc)
+            for port in coordinator.data.values()
+            for desc in SWITCH_PORT_DETAILS_SWITCHES
+            if desc.exists_func(switch, port)
+        )
 
     gateway_coordinator = await controller.get_gateway_coordinator()
     if gateway_coordinator:
         for gateway in gateway_coordinator.data.values():
             entities.extend(
-                OmadaGatewayPortSwitchEntity(
-                    gateway_coordinator, gateway, p.port_number, desc
-                )
+                OmadaDevicePortSwitchEntity[
+                    OmadaGatewayCoordinator, OmadaGateway, OmadaGatewayPortStatus
+                ](gateway_coordinator, gateway, p.port_number, str(p.port_number), desc)
                 for p in gateway.port_status
-                for desc in GATEWAY_PORT_SWITCHES
-                if desc.exists_func(p)
+                for desc in GATEWAY_PORT_STATUS_SWITCHES
+                if desc.exists_func(gateway, p)
+            )
+            entities.extend(
+                OmadaDevicePortSwitchEntity[
+                    OmadaGatewayCoordinator, OmadaGateway, OmadaGatewayPortConfig
+                ](gateway_coordinator, gateway, p.port_number, str(p.port_number), desc)
```

Comment:
```suggestion
                ](gateway_coordinator, gateway, p.port_number, desc)
```

Reply:
Yeah, I think this is a bit tidier. I'll commit these suggestions.